### DISCUSSION TITLE
Example of applying xmap on serial

### DISCUSSION THREAD

**Discussion by mattiasmar** [jax-ml/jax#8649] [Q&A] [answered]
Looking for an example applying xmap on the elements of a stax.serial
Thanks!

**Comment by hawkinsp** [comment]
This doesn't answer your question, but please note: we discourage the use of `stax` for most users who are looking for a neural network library. In general, we point users to Flax or Haiku.

Can you say more about what you are trying to do? I can't entirely parse the question.

**Reply by mattiasmar** [reply]
I want to build a (simple) multi layer workload where I can control the parallelism (mp, dp or none) on a per layer basis. stax is not a requirement, I only thought it would useful for my task. Thanks for pointing me to Flax and Haiku.

**Reply by mattiasmar** [reply]
I did some reading on Flax and the linen API.
To my understanding from the documentation the parallelization granted by pjit and xmap cannot be achieved with the linen package (only the vectorization of vmap is supported). Can you confirm this limitation?
Source: https://flax.readthedocs.io/en/latest/design_notes/lift.html?highlight=pmap#supported-transformations

**Reply by avital** [reply]
Hi @mattiasmar -- we definitely use `pjit` in Flax codebases, but typically only at the top-level training code. The lifted transformations you just linked to talk about using JAX transformations within modules (which requires special-casing because there are now new implicit inputs and outputs to the functions -- the variable collections such as parameters). 

If you want to use `pjit` at the top level with Flax you just wrap your `Module.apply` method (which is always a pure function). Haiku would behave similarly.

https://github.com/google-research/t5x is an example codebase that uses Flax and `pjit`. I think @marcvanzee is working on a smaller educational example (Marc, is it ready yet?)

That being said, sounds like you really do want to control parallelism on each layer separately? Could you explain more? Do you just want to decide on which parameters would be sharded and which would be replicated? You could do that at the top level with pjit. If you explain more about your request we could see whether our example covers your usecase.

**Reply by mattiasmar** [reply]
Hi @avital,

Ideally I would like to see an example that can import *any* JAX model, and assign automatically a unique (e.g. random) pjit spec for each layer. My goal is then to extend that example with a search for the optimal pjit specs for the provided model.

[FlexFlow](https://github.com/flexflow/FlexFlow) demonstrated using MCMC for searching for the optimal hybrid model-data parallelism, even controlling the degree of parallelism for each layer. Do you think one can achieve that with JAX?

**Reply by marcvanzee** [reply]
@mattiasmar sounds like a ambitious goal! If you are looking for a simple example for using `pjit`, I made one last week: https://github.com/marcvanzee/flax/tree/pjit-example/examples/siren

I still have to include a README, but the notebook already contains a lot of explanation.

**Reply by mattiasmar** [reply]
@marcvanzee Thanks! Is there anything I can do in order to run your notebook on a node with GPUs (rather than TPU)?
I get the error: `AttributeError: 'jaxlib.xla_extension.GpuDevice' object has no attribute 'coords'` from the method  `bounds_from_last_device`

Also: What requirement do you think primarily makes the goal ambitious: The application on *any* JAX model or the creating pjit spec for each layer?

**Reply by marcvanzee** [reply]
Ah, it says in the notebook it should be ran on a Cloud TPU v3-8, but I should probably write that down more clearly! I think the `coords` attribute is only there when using TPUs, but I am not sure which version this is for. @skye, I think you probably know more about this?

**Reply by mattiasmar** [reply]
The comment about the dependency on TPU in the notebook is clear. I was asking if there is a workaround.

**Reply by marcvanzee** [reply]
I am not sure about that actually, maybe @skye or @levskaya know this?

**Reply by mattiasmar** [reply]
FYI: Also on a collab with TPU runtime the `coords `attribute is missing: `AttributeError: 'jaxlib.xla_extension.Device' object has no attribute 'coords'`

**Reply by marcvanzee** [reply]
The Colabs are unfortunately still using outdated TPUs :-(.
This is why I made the example as a Jupyter notebook so you can run it on a Cloud TPU v3-8.
Sorry about that, we really hope Colab will soon upgrade!

**Reply by mattiasmar** [reply]
I created the MLP example below in an attempt to control the pjit of individual layers. Could you please help me define the tuple of PartitionSpec's of "Example 2"?
```
import jax
import jax.numpy as jnp
from jax import random
from jax.experimental.pjit import pjit
from jax.experimental import PartitionSpec, maps
from jax.scipy.special import logsumexp
import numpy as np
# A helper function to randomly initialize weights and biases
# for a dense neural network layer
def random_layer_params(n_in, n_out, key, scale=1e-2):
    w_key, b_key = random.split(key)
    return scale * random.normal(w_key, (n_in, n_out)), scale * random.normal(b_key, (n_out,))

# Initialize all layers for a fully-connected neural network with sizes "sizes"
def init_network_params(sizes, key):
    keys = random.split(key, len(sizes))
    return [random_layer_params(n_in, n_out, k) for n_in, n_out, k in zip(sizes[:-1], sizes[1:], keys)]

def relu(x):
    return jnp.maximum(0,x)

params = init_network_params([8,16,32,10], random.PRNGKey(0))
def predict(params, image):
  # per-example predictions
    activations = image
    for w, b in params[:-1]:
        outputs = jnp.dot(activations,w) + b
        activations = relu(outputs)
  
    final_w, final_b = params[-1]
    logits = jnp.dot(activations,final_w ) + final_b
    return logits - logsumexp(logits)
input = np.repeat(np.random.randint(0,9,(1,8)),repeats=4,axis=0)
mesh = maps.Mesh(np.array(jax.devices()).reshape(2,2), ('x','y'))

v,t = jax.tree_flatten(params)
print(t) 
#=========
# PyTreeDef([(*, *), (*, *), (*, *)])
#=========

#==========Example 1: WORKS=============
print("Basic pjit")
f = pjit( predict,in_axis_resources=PartitionSpec('x'),out_axis_resources=None)
with maps.mesh(mesh.devices, mesh.axis_names):
    out = f(params,input)



#==========Example 2:  DOESN'T WORK=============
print("Per layer pjit")
f = pjit( predict,in_axis_resources=(PartitionSpec('x'),PartitionSpec('x'),PartitionSpec('x')), out_axis_resources=None)
with maps.mesh(mesh.devices, mesh.axis_names):
    out = f(params,input)

#ValueError: pjit in_axis_resources specification must be a tree prefix of the corresponding value, 
#got specification (PartitionSpec('x',), PartitionSpec('x',), PartitionSpec('x',)) for value tree 
#PyTreeDef(([(*, *), (*, *), (*, *)], *)). Note that pjit in_axis_resources that are non-trivial 
#pytrees should always be wrapped in a tuple representing the argument list.

```

**Reply by skye** [reply]
`GpuDevice`s never have a `coords` field, because they're generally not arranged in a mesh topology, so this behavior is expected. I believe the coords are used in Marc's example to generate a topology-aware device mesh, so as a workaround, you could add something like this to the top of `get_mesh` in siren.py:
```python
if jax.default_backend() == 'gpu':
  return maps.Mesh(np.array(jax.devices()).reshape(2,2), ('data','model'))
assert jax.default_backend() == 'tpu'
```

**Reply by skye** [reply]
For your example 2 that doesn't work, you're providing 3 PartitionSpecs, but your function f only takes two arguments.

**Reply by mattiasmar** [reply]
@skye Thanks. The intention behind example 2 was that the 3 PartitionSpecs would match the 3 layers listed by `tree_flatten`: (`PyTreeDef([(*, *), (*, *), (*, *)])`) and eventually I want to give a different PartionSpec for each layer. From what I can tell from the documentation the argument `in_axis_resources` and `out_axis_resources` of `pjit` must match (or be broadcasted) to the PyTreeDef of the input fun to pjit (in our case the argument `params` of method `predict`).

Could you tell me what I got wrong in my understanding (or could you help me achieve this layer wise control of PartionSpec  that I'm trying to achieve)?

**Reply by mattiasmar** [reply]
If I decorate each of the layers inside the `predict` function with `pjit` (and accompany each such decoration with a `PartionSpec` at the same time) it works (obviously).
However, my goal is to decorate all of the layers of the model (ideally any model) in a post-processing step. This is in contrast to surgically entering the code and adding calls to pjit in attachment to every layer (which can be prohibitively complicated to do). 

So if I somehow can give a list of PartionSpec's (or a tuple/dict/PyTree of PartionSpec's) that relates to all the layers of model in a single pjit call, then I would have achieved my goal. Is this possible? Alternatively is this possible with xmap?

**Comment by mattiasmar** [comment]
I an attempt to be even more specific, and to communicate to you better, I defined the `params` object as a dictionary with keys (`"weights"` and `"biases"`)
Inside the code I added the print outs, so you don't actually need to run the code.


```
import jax
import jax.numpy as jnp
from jax import random, pmap
from jax.experimental.pjit import pjit
from jax.experimental import PartitionSpec, maps
from jax.scipy.special import logsumexp
import numpy as np
import os

os.environ['NCCL_LAUNCH_MODE'] = "PARALLEL"

# A helper function to randomly initialize weights and biases
# for a dense neural network layer
def random_layer_params(n_in, n_out, key, scale=1e-2):
    w_key, b_key = random.split(key)
    return {"weights": scale * random.normal(w_key, (n_in, n_out)), "biases": scale * random.normal(b_key, (n_out,))}


# Initialize all layers for a fully-connected neural network with sizes "sizes"
def init_network_params(sizes, key):
    keys = random.split(key, len(sizes))
    return [random_layer_params(n_in, n_out, k) for n_in, n_out, k in zip(sizes[:-1], sizes[1:], keys)]

def relu(x):
    return jnp.maximum(0, x)

layers = [28 * 28,512,10]
params = init_network_params(layers, random.PRNGKey(0))

def f(activations, layer):
    outputs = jnp.dot(activations, layer['weights']) + layer['biases']
    activations = relu(outputs)
    return activations


def predict(params, image):
    # per-example predictions
    activations = image
    for i, layer in enumerate(params[:-1]):
        activations = f(activations, layer)

    logits = f(activations, params[-1])
    return logits - logsumexp(logits)


input = np.repeat(np.random.randint(0, 9, (1, 28 * 28)), repeats=8 * 256, axis=0)
mesh = maps.Mesh(np.array(jax.devices()).reshape(2, 4), ('x', 'y'))

v, t = jax.tree_flatten(params)
print("Structure of 'params'=",t)
#=========
#Structure of 'params'= PyTreeDef([{'biases': *, 'weights': *}, {'biases': *, 'weights': *}])
#=========

specs= list()
for _ in range(len(layers)-1):
  specs.append({"biases":PartitionSpec('x'),"weights":PartitionSpec('x')})
specs.append(PartitionSpec(None)) #NOTE! Both with and without this line the error message remains the same.
specs = tuple(specs)
print("Structure of 'PartitionSpec Tree'=", specs)
#==========
# Structure of 'PartitionSpec Tree'= ({'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, {'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, PartitionSpec(None,))
#==========

f1 = pjit( predict,in_axis_resources=specs, out_axis_resources=None)
#=======
# FAILURE:
# ValueError: pjit in_axis_resources specification must be a tree prefix of the corresponding value,
#  got specification         ({'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, {'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, PartitionSpec(None,))
#  for value tree PyTreeDef(([{'biases': *,                   'weights':                   *}, {'biases': *,                   'weights':                  *}],                   *)).
#  Note that pjit in_axis_resources that are non-trivial pytrees should always be wrapped in a tuple representing the argument list.
#=======
with maps.mesh(mesh.devices, mesh.axis_names):
     out = f1(params,input)
print(out.shape)
```

**Comment by mattiasmar** [comment, marked as answer]
OK! Now I got it right! :smile:
This is the story: Under the hood the jax `wraps` method (of utils.py) adds an additional element to the PyTreeDef. When accounting for that my code runs without errors (yet to determine if the math is preserved too).
Thanks @skye @avital @hawkinsp @marcvanzee

**Reply by skye** [reply]
Hi @mattiasmar , I'm not sure what you mean by `wraps` adding an additional elements. In your new example, you should indeed add the extra `PartitionSpec(None)` for the `image` argument of `predict`. However, you need to put all the layer PartitionSpecs in a list, so it matches the list structure of the `params` argument. Here's a cleaned up version of the error message where you can see that the specification is the same as the value PyTreeDef except for the missing list:
```
got specification (
{'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, 
{'biases': PartitionSpec('x',), 'weights': PartitionSpec('x',)}, 
PartitionSpec(None,)
)

for value tree PyTreeDef( (
[{'biases': *, 'weights': *}, 
 {'biases': *, 'weights': *}], 
*)
).
```
I think the following code should work:
```python
layer_specs= list()
for _ in range(len(layers)-1):
  layer_specs.append({"biases":PartitionSpec('x'),"weights":PartitionSpec('x')})
image_spec = PartitionSpec(None)
specs = (layer_specs, image_spec)
print("Structure of 'PartitionSpec Tree'=", specs)
```

**Reply by mattiasmar** [reply]
Clearly the second PartitionSpec is needed for "image" argument of predict. Thanks for pointing that out @skye.
